nemo/collections/nlp/data/machine_translation/preproc_mt_data.py

```diff
@@ -0,0 +1,406 @@
+
+                tmp_f_src.close()
+                tmp_f_tgt.close()
+                tar_file_ptr, global_batch_ctr, num_files_in_tar, tar_file_ctr = self.write_batches_to_tarfiles(
+                    out_dir=out_dir,
+                    num_batches_per_tarfile=num_batches_per_tarfile,
+                    clean=clean,
+                    max_seq_length=max_seq_length,
+                    min_seq_length=min_seq_length,
+                    src_fname=tmp_f_src.name,
+                    tgt_fname=tmp_f_tgt.name,
+                    num_tokens=tokens_in_batch,
+                    encoder_tokenizer=encoder_tokenizer,
+                    decoder_tokenizer=decoder_tokenizer,
+                    num_files_in_tar=num_files_in_tar,
+                    tar_file_ptr=tar_file_ptr,
+                    tar_file_ctr=tar_file_ctr,
+                    global_batch_ctr=global_batch_ctr,
+                )
+                tar_file_ptr.close()
+                os.remove(tmp_f_src.name)
+                os.remove(tmp_f_tgt.name)
+
+                if num_files_in_tar != num_batches_per_tarfile:
+                    os.remove(os.path.join(out_dir, 'batches.tokens.%d.%d.tar' % (tokens_in_batch, tar_file_ctr)))
+                    global_batch_ctr -= num_files_in_tar
+                    print('Dropping %d batches because of overflow' % (num_files_in_tar))
+
+                json.dump({'num_batches': global_batch_ctr}, open(metadata_path, 'w'))
+
+        tar_file_paths = glob.glob(f'{out_dir}/batches.tokens.{tokens_in_batch}.*.tar')
+
+        return tar_file_paths, metadata_path
+
+    def train_tokenizers(
+        self,
+        out_dir,
+        src_fname,
+        tgt_fname,
+        shared_tokenizer,
+        encoder_tokenizer_name,
+        encoder_tokenizer_vocab_size,
+        decoder_tokenizer_name,
+        decoder_tokenizer_vocab_size,
+    ):
+        os.makedirs(out_dir, exist_ok=True)
+
+        if encoder_tokenizer_name != 'yttm' or decoder_tokenizer_name != 'yttm':
+            raise NotImplementedError(f"Currently we only support yttm tokenizer.")
+
+        if shared_tokenizer:
+            encoder_tokenizer_model = os.path.join(
+                out_dir, 'shared_tokenizer.%d.BPE.model' % (encoder_tokenizer_vocab_size)
+            )
+            decoder_tokenizer_model = encoder_tokenizer_model
+            if self.global_rank == 0:
+                if os.path.isfile(encoder_tokenizer_model):
+                    logging.info(
+                        f'Shared tokenizer model {encoder_tokenizer_model} already exists. Remove file if training a new tokenizer model.'
+                    )
+                else:
+                    logging.info(
+                        f'Shared tokenizer model {encoder_tokenizer_model} not found. Training tokenizer model.'
+                    )
+                    os.system('cat %s %s > %s' % (src_fname, tgt_fname, '/tmp/concat_dataset.txt'))
```

Comment:
please use `tempfile.TemporaryDirectory` as context instead of `/tmp/`

Reply:
Done.